### DISCUSSION TITLE
Optional API operations

### DISCUSSION THREAD

**Discussion by ways** [schemathesis/schemathesis#2440] [Q&A] [answered]
Hi

I'm testing schematesis for writing an open source EDR validator (https://ogcapi.ogc.org/edr/)

Some APIs will have an optional endpoint /collections/<name>/locations. Here's some example code of how I think that could work:

```
def test_valid_locations(case):
   pass

@schema.parametrize()
def test_api(case):
    if case.path.endswith("/locations"):
        test_valid_locations(case)
    else:
        ''' Run schemathesis standard tests '''
        case.call_and_validate()
```

Pytest complains when this path doesn't exist in the schema:

```
__________________ ERROR collecting use_hooks_to_find_extents,_collections.py ___________________
/home/larsfp/src/edr-profil/venv/lib/python3.10/site-packages/schemathesis/extra/pytest_plugin.py:219: in collect
    ???
venv/lib/python3.10/site-packages/schemathesis/utils.py:43: in fail_on_no_matches
    pytest.fail(f"Test function {node_id} does not match any API operations and therefore has no effect")
E   Failed: Test function use_hooks_to_find_extents,_collections.py::test_valid_locations does not match any API operations and therefore has no effect
```

I've also attempted to use test_valid_locations with: `@schema.include(path_regex="/locations$").parametrize()`

Am I approaching this the wrong way?

**Comment by Stranger6667** [comment]
Hi, the code sample looks a bit incomplete, as there is no `@schema.parametrize()` on `test_valid_locations`, but `pytest` fails on it, which only happens when that decorator is present. 

Could you, please, share the complete version of the sample including how you define the `schema` variable + the API schema?

**Comment by ways** [comment]
Here's the complete code: https://gist.github.com/ways/77ebefb2740178d44ec8b45cc8c6d2ba
And an API to test on: https://edrisobaric.k8s.met.no/
To run it you need to add these parameters: --openapi https://edrisobaric.k8s.met.no/api --url https://edrisobaric.k8s.met.no

**Comment by ways** [comment, marked as answer]
Never mind. I solved it by using @schema.parametrize() instead of @schema.include(path_regex="/position") and just returning if the path is wrong:

```
@schema.parametrize()
def test_positions(case):
    if not case.path.endswith("/position"):
        return

    response = case.call()
...
```
